src/math/mod.rs

```diff
@@ -29,6 +29,124 @@ pub fn canon_egcd(a: i64, b: i64, c: i64) -> Option<(i64, i64, i64)> {
     }
 }
 
+fn pos_mod(n: i64, m: i64) -> i64 {
+    if n < 0 {
+        n + m
+    } else {
+        n
+    }
+}
+
+fn mod_mul(a: i64, b: i64, m: i64) -> i64 {
+    pos_mod((a as i128 * b as i128 % m as i128) as i64, m)
+}
+
+/// Assuming m >= 2 and exp >= 0, finds base ^ exp % m in logarithmic time
+pub fn mod_exp(mut base: i64, mut exp: i64, m: i64) -> i64 {
+    assert!(m >= 2);
+    assert!(exp >= 0);
+    let mut ans = 1 % m;
+    base = base % m;
+    while exp > 0 {
+        if exp % 2 == 1 {
+            ans = mod_mul(ans, base, m);
+        }
+        base = mod_mul(base, base, m);
+        exp /= 2;
+    }
+    pos_mod(ans, m)
+}
+
+/// Assuming m >= 2, finds multiplicative inverse of n under modulus m
+pub fn mod_inv(n: i64, m: i64) -> i64 {
+    mod_exp(n, m - 2, m)
+}
+
+fn miller_test(n: i64, d: i64, r: i64, a: i64) -> bool {
+    let mut x = mod_exp(a, d, n);
+    if x == 1 || x == n - 1 {
+        return true;
+    }
+    for _ in 0..r {
+        x = mod_mul(x, x, n);
+        if x == n - 1 {
+            return true;
+        }
+    }
+    false
+}
+
+const BASES: [i64; 12] = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37];
+/// Assuming x >= 0, returns true if x is prime
+pub fn is_prime(n: i64) -> bool {
+    assert!(n >= 0);
+    if n <= 1 {
+        return false;
+    }
+    if n <= 3 {
+        return true;
+    }
+    let mut d = n - 1;
+    let mut r = 0;
+    while d % 2 == 0 {
+        d /= 2;
+        r += 1;
+    }
+    for base in BASES.iter() {
+        if *base <= n - 2 && !miller_test(n, d, r, *base) {
+            return false;
+        }
+    }
+    true
+}
+
+fn pollard_rho(n: i64) -> i64 {
+    for a in 1..n {
+        let f = |x| pos_mod(mod_mul(x, x, n) + a, n);
+        for b in 0..n {
+            let mut x = b;
+            let mut y = b;
+            loop {
+                x = f(x);
+                y = f(f(y));
+                let p = num::fast_gcd((x - y).abs(), n);
+                if p > 1 && p < n {
+                    return p;
+                }
+                if x == y {
+                    break;
+                }
+            }
+        }
+    }
+    panic!("No divisor found!");
+}
+
+/// Assuming x >= 0, finds the prime factorization of n
+pub fn factorize(n: i64) -> Vec<i64> {
+    assert!(n >= 0);
+    let mut factors = Vec::new();
+    if n < 2 {
+        return factors;
+    }
+    let mut stack = Vec::new();
+    stack.push(n);
+    while let Some(top) = stack.pop() {
+        if is_prime(top) {
+            factors.push(top);
+        } else if top % 2 == 0 {
+            stack.push(2);
```

Comment:
You could directly do `factors.push(2)`. Btw, are you avoiding recursion for efficiency reasons?

Reply:
Efficiency + I think it is cleaner to implement this by iteration, at least I wasn't able to come up with a better solution using recursion.

Reply:
Yeah it's nice! Hm since we're going for efficiency, wouldn't it make sense to deal with the even case before testing for primality?

Reply:
It makes it a bit less clean because then you have to first test for 2, then test for other evens, then test for whether its a prime and finally use pollard_rho if none of those are true (overall, it adds another branch). I could add this check into the is_prime function itself though, so the cost of calling is_prime on something that is even is smaller.

Reply:
Hm we could also remove the powers of two immediately, if you prefer:

```
let r = n.trailing_zeros();
let mut factors = vec![2; r];
let mut stack = vec![n >> r];
```

Reply:
I added this, but I had to add an if statement after to remove the top element if it is 1, because n >> r can be 1.

Reply:
I swapped the if statement out for a match, I think it looks a bit better now.